### ISSUE TITLE
zero-shot accuracies on okvqa and gqa seems to be much lower than paper reported using blip2_caption_opt2.7b

### ISSUE THREAD

**Issue by Laeglaur** [salesforce/LAVIS#188]
Hi, since this repo currently doesn't release blip2_opt test script on VQA task, I have writed test scipt according to paper and `predict_answers` in blip2_t5.py 
Here is my code.
```    
def predict_answers(
        self,
        samples,
        num_beams=5,
        inference_method="generate",
        use_nucleus_sampling=False,
        max_len=10,
        min_len=1,
        num_ans_candidates=128,
        answer_list=None,
        prompt="",
        length_penalty=-1,
        repetition_penalty=1.0,
        **kwargs
    ):
        image = samples["image"]
        with torch.cuda.amp.autocast(enabled=(self.device != torch.device("cpu"))):
            image_embeds = self.ln_vision(self.visual_encoder(image))
        image_atts = torch.ones(image_embeds.size()[:-1], dtype=torch.long).to(
            image.device
        )
        query_tokens = self.query_tokens.expand(image_embeds.shape[0], -1, -1)
        query_output = self.Qformer.bert(
            query_embeds=query_tokens,
            encoder_hidden_states=image_embeds,
            encoder_attention_mask=image_atts,
            return_dict=True,
        )

        inputs_opt = self.opt_proj(query_output.last_hidden_state)
        atts_opt = torch.ones(inputs_opt.size()[:-1], dtype=torch.long).to(image.device)

        if isinstance(samples["text_input"], str):
            samples["text_input"] = [samples["text_input"]]
        if prompt:
            text_input = [prompt.format(question) for question in samples["text_input"]]
        else:
            text_input = samples["text_input"]

        input_tokens = self.opt_tokenizer(text_input, padding="longest", return_tensors="pt").to(image.device)
        input_ids = input_tokens.input_ids
        encoder_atts = torch.cat([atts_opt, input_tokens.attention_mask], dim=1)

        if use_nucleus_sampling:
            input_embeds = inputs_opt.repeat_interleave(1, dim=0)
            num_beams = 1
        else:
            input_embeds = inputs_opt.repeat_interleave(num_beams, dim=0)

        device_type = "cuda" if "cuda" in str(self.device) else "cpu"
        with torch.amp.autocast(device_type=device_type, dtype=torch.bfloat16):
            outputs = self.opt_model.generate(
                input_ids=input_ids,
                query_embeds=input_embeds,
                attention_mask=encoder_atts,
                do_sample=use_nucleus_sampling,
                num_beams=num_beams,
                max_new_tokens=max_len,
                min_length=min_len,
                length_penalty=length_penalty,
               # eos_token_id=self.eos_token_id,

            )

            prompt_length = input_tokens.input_ids.shape[1]
            output_text = self.opt_tokenizer.batch_decode(outputs[:, prompt_length:], skip_special_tokens=True)
            output_text = [text.strip().split('\n')[0].strip(string.punctuation) for text in output_text] # using first strip to remove first eos.
        # add
        if self._apply_lemmatizer:
            output_text = self._lemmatize(output_text)
        return output_text
```
Test Config
```
max_len=20, 
min_len=1, 
beam_search=5, 
use_nucleus_sampling=False, 
length_penalty=-1, 
_apply_lemmatizer= True, 
prompt = "Question: {}, Answer:"
```
- Using max_len=20 because the accuracy is slightly better than using max_len=10 which is Flan-xl-t5's max_len.
- max_len=20: 17.80, max_len=10: 17.37

And I use official released pretrain blip2_opt weight (https://storage.googleapis.com/sfr-vision-language-research/LAVIS/models/BLIP2/blip2_caption_opt2.7b.pth) to test on OKVQA and GQA.

Here is the test accuracy
Source | LLM | Vision Model | OK-VQA | GQA
-- | -- | -- | -- | --
paper | Flan-xl-t5 | ViT-G | 40.7 | 44.2
test by myself | Flan-xl-t5 | ViT-G | 41.13 | 43.96
paper | opt2.7b | ViT-G |  31.7 | 34.6
test by myself | opt2.7b | ViT-G | 17.80 | 14.406

* NOTE: In order to double check datasets and enviroment are fine, I also test the two dataset accuracy of using flan-xl-t5 which have released test script and model. The accuracy seems to be align with paper reported.

And I found blip2_opt seems like predict `eos token` just after 'Answer:'.  So I can't use `eos_token_id=self.eos_token_id` like in `Blip2OPT.generate`, and need to use post-processing to remove first eos token.
Here is raw output.

> Question: what sport can you use this for? Answer:**\n**motorcycle\n\na motorcycle parked on', 'Question: what could this gentleman be carrying in that red bag? Answer: bag of money\n\na man in a bow', 'Question: what does this grow from? Answer: flower\n\nQuestion: what does this grow from', 'Question: what is the name of the items the hot dog are topped with? Answer:**\n\n**two hot dogs with toppings

**Comment by Laeglaur** [comment]
have solved above problem.
need to add `self.opt_tokenizer.padding_side = "left"`, because it will use the last token to decode result.
okvqa test result=31.62
```
def predict_answers(
        self,
        samples,
        num_beams=5,
        inference_method="generate",
        use_nucleus_sampling=False,
        max_len=10,
        min_len=1,
        num_ans_candidates=128,
        answer_list=None,
        prompt="",
        length_penalty=-1,
        repetition_penalty=1.0,
        **kwargs
    ):
        image = samples["image"]
        with torch.cuda.amp.autocast(enabled=(self.device != torch.device("cpu"))):
            image_embeds = self.ln_vision(self.visual_encoder(image))
        image_atts = torch.ones(image_embeds.size()[:-1], dtype=torch.long).to(
            image.device
        )
        query_tokens = self.query_tokens.expand(image_embeds.shape[0], -1, -1)
        query_output = self.Qformer.bert(
            query_embeds=query_tokens,
            encoder_hidden_states=image_embeds,
            encoder_attention_mask=image_atts,
            return_dict=True,
        )

        inputs_opt = self.opt_proj(query_output.last_hidden_state)
        atts_opt = torch.ones(inputs_opt.size()[:-1], dtype=torch.long).to(image.device)

        if isinstance(samples["text_input"], str):
            samples["text_input"] = [samples["text_input"]]
        if prompt:
            text_input = [prompt.format(question) for question in samples["text_input"]]
        else:
            text_input = samples["text_input"]
        self.opt_tokenizer.padding_side = "left"
        input_tokens = self.opt_tokenizer(text_input, padding="longest", return_tensors="pt").to(image.device)
        input_ids = input_tokens.input_ids
        encoder_atts = torch.cat([atts_opt, input_tokens.attention_mask], dim=1)

        if use_nucleus_sampling:
            input_embeds = inputs_opt.repeat_interleave(1, dim=0)
            num_beams = 1
        else:
            input_embeds = inputs_opt.repeat_interleave(num_beams, dim=0)

        device_type = "cuda" if "cuda" in str(self.device) else "cpu"
        with torch.amp.autocast(device_type=device_type, dtype=torch.bfloat16):
            outputs = self.opt_model.generate(
                input_ids=input_ids,
                query_embeds=input_embeds,
                attention_mask=encoder_atts,
                do_sample=use_nucleus_sampling,
                num_beams=num_beams,
                max_new_tokens=max_len,
                min_length=min_len,
                length_penalty=length_penalty,
               eos_token_id=self.eos_token_id,
            )

            prompt_length = input_tokens.input_ids.shape[1]
            output_text = self.opt_tokenizer.batch_decode(outputs[:, prompt_length:], skip_special_tokens=True)
            output_text = [text.strip() for text in output_text]
        # add
        if self._apply_lemmatizer:
            output_text = self._lemmatize(output_text)
        return output_text
```

**Comment by YuanLiuuuuuu** [comment]
I will be grateful if you can share me with your email. I want to consult you something about OPT zeroshot VQA. Thanks in advance!

**Comment by Laeglaur** [comment]
hi yuanliu, here is my email. (fish.cosmos@foxmail.com)

**Comment by renaissanceee** [comment]
Hi, do you know how to fintuning vqa on OPT model?

**Comment by dszpr** [comment]
Hi！ @Laeglaur 
I noticed that in forward function in blip2_opt.py, only the questions in the VQA dateset are used. Both the text_input and the target lables are derived from the opt_tokens
`


        text = [t + "\n" for t in samples["text_input"]]

        opt_tokens = self.opt_tokenizer(
            text,
            return_tensors="pt",
            padding="longest",
            truncation=True,
            max_length=self.max_txt_len,
        ).to(image.device)

        targets = opt_tokens.input_ids.masked_fill(
            opt_tokens.input_ids == self.opt_tokenizer.pad_token_id, -100
        )
        if self.prompt:
            targets[:, : self.prompt_length] = -100  # do not apply loss to the prompt

        empty_targets = (
            torch.ones(atts_opt.size(), dtype=torch.long).to(image.device).fill_(-100)
        )
        targets = torch.cat([empty_targets, targets], dim=1)

        inputs_embeds = self.opt_model.model.decoder.embed_tokens(opt_tokens.input_ids)
        inputs_embeds = torch.cat([inputs_opt, inputs_embeds], dim=1)
        attention_mask = torch.cat([atts_opt, opt_tokens.attention_mask], dim=1)

        with self.maybe_autocast():
            outputs = self.opt_model(
                inputs_embeds=inputs_embeds,
                attention_mask=attention_mask,
                return_dict=True,
                labels=targets,
            )
`
In the VQA task,  are the target lables supposed to be the answers in the VQA dataset? But the answers in the VQA datasetare not used in the blip2_opt.py. However, the answers are used as target lables in blip2_t5.py. It really confused.
Have you make any change to blip2_opt.py?
